test/es-module/test-esm-json.mjs

```diff
@@ -21,4 +24,40 @@ describe('ESM: importing JSON', () => {
     assert.strictEqual(code, 0);
     assert.strictEqual(signal, null);
   });
+
+  test('should load different modules when the URL is different', async (t) => {
+    const root = tmpdir.fileURL(`./test-esm-json-${Math.random()}/`);
+    try {
+      await mkdir(root, { recursive: true });
+
+      await t.test('json', async () => {
+        const url = new URL('./foo.json', root);
+        await writeFile(url, JSON.stringify({ firstJson: true }));
+        const firstJson = await import(`${url}?a=1`, {
```

Comment:
We should also test it for hash, `${url}#a`/`${url}#b`.
Maybe `${url}?a=1`/`${url}?a=%31` as well (although it's not related to the fix).